### DISCUSSION TITLE
Zod validator calls refine and ignoring other validators

### DISCUSSION THREAD

**Discussion by hexwit** [colinhacks/zod#3060] [Q&A] [answered]
Hi, 
I need to validate presence of url, and that it has valid format. 
I know about url() validator, but I will need more fine grained checks.

So I tried to use refine().

```ts
validate: zodResolver(
    z.object({
        url: z.string()
            .min(1, "This is a required field")
            .refine((value) => {
                try {
                    new URL(value);
                    return true;
                } catch {
                    return false;
                }
            }, "Invalid URL")
    })
)
```

for some reason zod ignores '.min' validator and calls always refine(). 

Is this an issue or I just don't understand something? 

Thanks.

**Comment by JacobWeisenburger** [comment, marked as answer]
Zod tries to run all of the validation "chain", so that you would get as many of the errors as possible. Therefore, it will run a refine even if the prior checks failed.

**Reply by jedwards1211** [reply]
I think this intended behavior is more harmful than helpful though.  It makes sense to gather errors at other paths of an input data structure if one path fails, but not to gather more refinement failures on the same path.

`z.string().refine(s => ...)` doesn't try the refinement if the input isn't a `string`, so it's natural to expect that `z.string().url().refine(s => new URL(s).protocol === 'mqtts:')` won't try the refinement if the input isn't a url.

I think most users who run into this issue would prefer refinement chains to bail early?  Refinements seem like the exception to the rule -- afaik all other Zod nodes bail on the first node in a chain that fails parsing, except for refinements.

**Reply by jedwards1211** [reply]
The existing behavior can even violate type safety in refinement chains:

```ts
z.string()
  .refine((s): s is FooBarString => ...)
  .refine((s: FooBarString) => // no TS error, but this can get called with a non-FooBarString
    ...
  )
````

**Reply by scotttrinh** [reply]
> so it's natural to expect that `z.string().url().refine(s => new URL(s).protocol === 'mqtts:')` won't try the refinement if the input isn't a url.

But what about: `z.string().min(8).email()`? Should it check both conditions, or fail at whichever fails first? Proponents for "fail early" and "fail late" are equally vocal in their support of their preferred behavior here.

> The existing behavior can even violate type safety in refinement chains:

Indeed! We can _probably_ fix this, but might be easier said than done.

You can opt out of this by not using the built-in refinements (I personally skip them as well), and use `superRefine`, [which allows you to mark certain errors as fatal and otherwise gives you full control over the flow here.](https://zod.dev/?id=abort-early)

**Reply by jedwards1211** [reply]
I'm well-versed in Zod features, and I know how to work around this with `superRefine`, I would just rather not, because it's more cumbersome and less readable than better default behavior would be.  My argument is that being able to write `z.string().url().refine(s => new URL(s).protocol === 'mqtts:', 'must be an mqtts: url')` would be much more convenient for everyone overall.  We should always be open to questioning the status quo.

Basically, built-in checks should fail late, and refinements should fail early, and this difference is okay because built-in checks and refinements are actually very different.

It seems like the idea of trying multiple refinements probably came from the idea that we should get error messages from all the built-in checks that are violated, like `.min(1).max(80).email()`. 
 I agree this makes sense for built-in checks.  But note, built-in checks are not really refinements -- they don't chain together additional Zod nodes, change the output type, or use `ZodEffects` like refinements do, and they don't allow procedural code that could throw an error if any assumptions aren't met.

In most cases where a chain of Zod nodes is formed, parsing doesn't continue past a node that errored.  Refinements are additional nodes so they should conform to this by default, to be least surprising.  Built-in checks, on the other hand, are all applied by the same node, so it isn't surprising for them all to be tried.  But if any built-in checks fail, parsing shouldn't continue past that node to a subsequent `.pipe` or `.transform` or .refine` or whatever.  If people want it, there could be an option to opt in to successive refinements - but opt-in would be a more sensible default than opt-out.

**Reply by colinhacks** [reply]
I think the real answer here is to provide a mechanism for marking a refinement as "fatal". This will likely happen in Zod 4 but I don't have an RFC yet. 

```ts
z.string().url({ fatal: true })
z.string().refine(val => val.length > 10, { fatal: true })
```

Relatedly, refinements will be stored "in-schema" in Zod 4, as opposed to the current compositional approach with `ZodEffects`. There will be a standard format for both built-in checks and user-defined checks, and they'll live alongside each other in `._def.checks`. 

Ideally there would be a reasonable, easily-understandable rule for which checks are considered "default fatal" and which aren't. But practically speaking, I doubt it will be so simple. I see the argument for things like `.email()` and `.url()` being fatal but `.min()` and `.refine()` being non-fatal. We'll probably have to make case-by-case judgement calls for these.

**Reply by jedwards1211** [reply]
> `.refine()` being non-fatal

If there's a way we can disallow `.refine` predicates from returning `val is NewType` without `{ fatal: true }` then that would be best since otherwise it violates the input type expected by the next Zod node in the chain.  I guess at least we could make it not do the `predicate extends (val, ...) => val is infer O` logic without `fatal: true`.

**Reply by jedwards1211** [reply]
And if that's not possible, then maybe `.refine` could be fatal by default in Zod 4, and people who want it not to be would have to opt in?

**Reply by jedwards1211** [reply]
@colinhacks it might be nicer to have a method like `.fatal()` or `.bail()` or something that means stop refining if any of the previous refinements failed, so that we don't have to pass `{fatal: true}` to every one:

```ts
z.string().min(1).max(64).email().fatal().refine(...)
```

**Reply by nbibler** [reply]
Or, take an approach similar to `.optional()` which can be chained _or_ passed a schema. Conceptually, the following would do the same as the above, but in a more composable/obvious way:

```javascript
z.fatal(z.string().min(1).max(64).email()).refine(...)
```

**Reply by nbibler** [reply]
Actually, reflecting on this more, doesn't this desired behavior exist today with [`.pipe()`](https://zod.dev/?id=pipe)? I think the following would work as requested:

```javascript
z.string().min(1).max(64).email().pipe(z.string().refine(...))
```

**Comment by hexwit** [comment]
Is there any way to make it stop on the first failure?

**Comment by JacobWeisenburger** [comment]
not that I know of, but perhaps someone else would know.

**Comment by hexwit** [comment]
However this kind of validation works as expected: 
```
validate: zodResolver(
            z.object({
                email: z.string().email({message: "Email has invalid format"}).min(1, "This is a required field"),
            })
        ),
```
Chained validation, showed first appeared error. it is like something wrong with refine chaining.

**Reply by JacobWeisenburger** [reply]
Not sure what you are talking about, because this produces both errors.

```ts
const schema = z.object( {
    email: z.string().email( { message: "Email has invalid format" } ).min( 1, "This is a required field" ),
} )

const result = schema.safeParse( { email: '' } )
!result.success && console.log( result.error.issues )
// [
//     {
//         validation: "email",
//         code: "invalid_string",
//         message: "Email has invalid format",
//         path: [ "email" ]
//     }, {
//         code: "too_small",
//         minimum: 1,
//         type: "string",
//         inclusive: true,
//         exact: false,
//         message: "This is a required field",
//         path: [ "email" ]
//     }
// ]
```
